c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -482,6 +484,109 @@ class CachingAllocatorConfig {
   }
 };
 
+// create a custom memory plan
+class MemoryPlanner {
+ private:
+  std::vector<std::vector<AllocFreeEvent>> mem_plan;
+  std::vector<Block*> plan_blk_ptr;
+  std::vector<unsigned int> plan_counter;
+  std::vector<std::map<intptr_t, bool>> allocated_blocks;
+  std::vector<size_t> planning_size;
+  std::vector<unsigned int> non_rec_length;
+
+ public:
+  ska::flat_hash_map<void*, Block*> assigned_planned_blocks;
+  void create_plan(
+      std::vector<std::vector<AllocFreeEvent>> plan,
+      std::vector<Block*> plan_ptrs,
+      std::vector<size_t> plan_size,
+      std::vector<unsigned int> non_rec_lengths) {
+    mem_plan = plan;
+    plan_blk_ptr.resize(plan.size(), nullptr);
+    plan_counter.resize(plan.size(), 1);
+    planning_size.resize(plan.size(), 0);
+    allocated_blocks.resize(plan.size());
+    non_rec_length.resize(plan.size(), 0);
+    plan_blk_ptr = plan_ptrs;
+    planning_size = plan_size;
+    non_rec_length = non_rec_lengths;
+  }
+
+  Block* malloc(size_t size, int device, cudaStream_t stream) {
+    Block* block = nullptr;
+    if ((plan_blk_ptr[device] != nullptr) &&
+        (mem_plan[device][plan_counter[device]].size > 0) &&
+        (size == (unsigned int)mem_plan[device][plan_counter[device]].size)) {
+      intptr_t planned_ptr = mem_plan[device][plan_counter[device]].ptr;
+
+      auto it_ptr_lower_bound =
+          allocated_blocks[device].lower_bound(planned_ptr);
+      if ((planned_ptr + size <= planning_size[device]) &&
+          (it_ptr_lower_bound == allocated_blocks[device].end()) ||
+
+           it_ptr_lower_bound->second &&
+
+            planned_ptr+ s ize <= (unsigned int) t_ptr_lower_bound->first)))) {
+        block = new Block(
+            device,
+            stream,
+            size,
+            nullptr,
+            reinterpret_cast<void*>(
+                reinterpret_cast<intptr_t>(plan_blk_ptr[device]->ptr) +
+                planned_ptr));
+        block->planned = true;
+        allocated_blocks[device][planned_ptr] = true;
+        allocated_blocks[device][planned_ptr + size - 1] = false;
+      }
+      plan_counter[device] += 1;
+      if (plan_counter[device] == mem_plan[device].size()) {
+        plan_counter[device] = non_rec_length[device]+ 1 ;
+      }
+    }
+    return block;
```

Comment:
I guess `block == nullptr` means planned allocation failed? If so, please add comments to highlight that.

Reply:
Done. Thank you.